src/core/thread/network_diagnostic.cpp

```diff
@@ -521,6 +521,12 @@ void NetworkDiagnostic::HandleDiagnosticGetRequest(Coap::Header &aHeader, Messag
 
     SuccessOrExit(error = FillRequestedTlvs(aMessage, *message, networkDiagnosticTlv));
 
+    if (message->GetLength() == header.GetLength() && networkDiagnosticTlv.GetLength() > 0)
```

Comment:
I think it would be simpler to always call `SetPayloadMarker()` (i.e. remove the if check at L515) above and then remove the `networkDiagnosticTlv.GetLength()` check here.